server/server_test.go

```diff
@@ -39,6 +40,82 @@ func (s *safeWriter) Write(p []byte) (n int, err error) {
 	return s.writer.Write(p)
 }
 
+func TestTLSHandshakeErrorWriter(t *testing.T) {
+	var mutex sync.Mutex
```

Comment:
I don't see that you're using [`t.Parallel()`](https://pkg.go.dev/testing#T.Parallel) anywhere, nor are you starting any goroutines, so why would you need to add a mutex here to protect against concurrency?